### ISSUE TITLE
Various loo/waic items

### ISSUE THREAD

**Issue by jgabry** [paul-buerkner/brms#414]
A few things related to loo/waic: 

* The recent [question on discourse](http://discourse.mc-stan.org/t/waic-and-loo-in-multivariate-models-with-different-response-variables/4007) about waic and loo when models have different outcomes reminded me to recommend that **brms** check whether the response variables are the same before allowing model comparison. In rstanarm we do this by computing the hash of the outcome and saving it as an attribute in the object returned by the loo method for stanreg objects 

    https://github.com/stan-dev/rstanarm/blob/f11d6bf92b6c78ce64896caf78c2237ebd015abb/R/loo.R#L802-L813

    and then when `compare_models` is called we check that the hashes are the same

    https://github.com/stan-dev/rstanarm/blob/f11d6bf92b6c78ce64896caf78c2237ebd015abb/R/loo.R#L850-L856

* The loo and waic methods in brms are called `LOO` and `WAIC`. Any reason not to define methods for the `loo` and `waic` generics from the __loo__ package? If you're open to that then the `LOO` and `WAIC` methods could be deprecated so as not to break backwards compatibility. 

* There is no generic in __loo__ or __rstantools__ for comparing models but there probably should be so that __brms__, __rstanarm__, and other packages can use the same function name rather than `compare_ic` and `compare_models`. Would you be open to using that generic if it existed? 

Anyway, I'm happy to submit PRs for these things if you're open to them.

**Comment by paul-buerkner** [comment]
Thanks for bringing these points up.

- brms already checks whether response variables are equal, but only issues a warning when they are not.  I may change that to an error if you think that's better.  See function `match_response` in https://github.com/paul-buerkner/brms/blob/master/R/loo-helpers.R. 

- brms also supports `loo` and `waic`, so both upper and lower case work. We could deprecate the upper case once, but I don't think having both is really problematic.

- I would definitely like to have a common generic for model comparisons whatever the name is.

**Comment by jgabry** [comment]
I should have numbered the points so it would be easier to respond now. I'll pretend they're numbered 1-3. 

So regarding (1) I tried comparing the following simple models that have different outcome variables but I didn't get the warning: 

```
# packageVersion("brms") is 2.2.1
fit1 <- brm(mpg ~ wt, data = mtcars)
fit2 <- brm(wt ~ mpg, data = mtcars)

l1 <- loo(fit1)
l2 <- loo(fit2)
compare_ic(l1, l2)
```
Output with no warning:
```
             LOOIC   SE
fit1        167.04 8.51
fit2         50.96 9.45
fit1 - fit2 116.09 7.42
```
That's why I assumed it wasn't checking, but I do see `match_response`. Is this somehow not getting caught by `match_response`?

Regarding (2), the reason I brought it up is that I keep seeing people using the capitalized versions `LOO` and `WAIC` and, while certainly not a serious problem at all, I do think it's a bit odd/confusing to have both. So I would definitely be in favor of deprecating the capitalized names even though the lowercase names also work, but not a big deal either way! 

And (3), ok great!

**Comment by paul-buerkner** [comment]
1) When running `LOO(fit1, fit2)` the warning occurs. I just realize that `match_resppnse` is not called when running `compare_ic`. I will need to find a way to make this possible. Maybe hashing is indeed a good option here. Let me think about this.

2) Ok.

3) Actually, I am not sure if `compare_models` is a good name, because the function takes information criterion values rather than models. What do you think?

**Comment by paul-buerkner** [comment]
While we are at it. Can we also define a `kfold` generic somewhere?

**Comment by jgabry** [comment]
Yes! We should definitely have a kfold generic too.

**Comment by paul-buerkner** [comment]
I like `loo_compare` personally. We already have `loo_model_weights` and the like.

**Comment by paul-buerkner** [comment]
Checking response values via hashing should now be working. 

I changed all the internal doc to now favor the lower case versions of `loo` and `waic` so that I expect the use of the upper case versions to decline. However, I decided not to deprecate `LOO` and `WAIC` for now as they are used in my own papers about brms and in many blog posts and case studies all over the internet.

**Comment by jgabry** [comment]
> Checking response values via hashing should now be working.

Nice!

> I changed all the internal doc to now favor the lower case versions of loo and waic so that I expect the use of the upper case versions to decline. However, I decided not to deprecate LOO and WAIC for now as they are used in my own papers about brms and in many blog posts and case studies all over the internet.

Ok cool, that makes sense.

**Comment by jgabry** [comment]
Ok so I guess we should add a loo_compare generic and default method to the
loo package and eventually deprecate loo::compare,
rstanarm::compare_models, and brms::compare_ic in favor of loo_compare
methods? The loo_compare methods could be used for waic and kfold
comparisons too.

**Comment by paul-buerkner** [comment]
Sounds good to me.

**Comment by jgabry** [comment]
Ok great. We’ll probably do a minor loo release every month for the next few months so I’ll get this in one of them.

**Comment by jgabry** [comment]
Did they ask you to submit an update by a certain date? For compare I’d say
we should do it soon but not rush because of CRAN if they want you to
submit this week. Either way, if you submit brms and then we do a loo
release before you can release brms again, I’ll wait to include the compare
update (and we won’t break anything in brms!).

Jonah

**Comment by paul-buerkner** [comment]
They gave me no deadline. I just don't like warnings on brms check results. ;-)

So it is not urgent really, just a question of whether its worth waiting one or two weeks for the loo stuff.

**Comment by paul-buerkner** [comment]
I have removed the ICs from the summary output on purpose as they are only
informative in comparison to the ICs of other models and in combination
with their standard error.

The recommended methods to compute / extract LOO and WAIC continue to be
loo() and waic() respectively.

2018-05-21 18:29 GMT+02:00 anyanova <notifications@github.com>:

**Comment by avehtari** [comment]
Related to unifying the outputs, the outputs of `loo`, `compare` and `compare_models` for rstanarm models are
```
> (loo1 <- loo(fit_1))

Computed from 4000 by 1187 log-likelihood matrix

         Estimate    SE
elpd_loo  -5178.0  74.3
p_loo        10.8   3.7
looic     10355.9 148.7
------
Monte Carlo SE of elpd_loo is 0.1.
```
Here elpd_loo is shown first, p_loo is on same scale and MCSE is for elpd_loo. looic is also shown although I would personally remove it as default, as it seems to confuse some users.
```
> compare(loo1,loo2)
elpd_diff        se 
     16.7       5.2 
> compare_models(loo1,loo2)

Model comparison: 
(negative 'elpd_diff' favors 1st model, positive favors 2nd) 

elpd_diff        se 
     16.7       5.2 
```
Only elpd_diff is shown. compare_models gives also informative message.
```
> compare(loo1,loo2,loo3)
     elpd_diff se_diff elpd_loo p_loo   looic  
loo3     0.0       0.0 -5161.0     12.2 10321.9
loo2    -0.3       1.0 -5161.3     11.4 10322.6
loo1   -17.0       5.0 -5178.0     10.8 10355.9
> compare_models(loo1,loo2,loo3)

Model comparison: 
(ordered by highest ELPD)

      elpd_diff se_diff elpd_loo p_loo   looic  
fit_3     0.0       0.0 -5161.0     12.2 10321.9
fit_2    -0.3       1.0 -5161.3     11.4 10322.6
fit_1   -17.0       5.0 -5178.0     10.8 10355.9
```
elpd_diff is in the first column and the order is based on elpd_loo. looic is given in the last column.

Note also, that all these report by default 1 decimal accuracy. Again, I would personally make looic optional (and not shown by default).

**Comment by paul-buerkner** [comment]
Do I understand it correctly, that the output structure varies between comparison of 2 models and 3 or more models? That seems somewhat strange to me.

I am happy to go more into the direction of `loo::compare` or `rstanarm::compare_models`, but I would like to keep the looic in the output because many users are used to it and I don't want to cause too much confusion by completely removing it.

**Comment by avehtari** [comment]
> Do I understand it correctly, that the output structure varies between comparison of 2 models and 3 or more models? That seems somewhat strange to me.

Yes, they are different, and that's why showed both output structures. I like the simplicity of two model comparison output, but I agree that it's strange that they look so different.

> but I would like to keep the looic in the output because many users are used to it and I don't want to cause too much confusion by completely removing it.

How about making it optional with looic=FALSE/TRUE option (with default to FALSE)? Now there are more columns, but with that option we could just multiply all other columns expect p_loo with -2.

**Comment by paul-buerkner** [comment]
The different look has the problem that they are inconsistent for the user not only visibly (that's fine) but structurally. This will be problematic for any packages building on top of these outputs.

I am fine with having a switch to turn it off and on. Assuming that in the end we have one `compare` generic and basic method in `loo` on which rstanarm, brms and others build their `compare` methods on, we could as well have the default of that switch be different across packages.

**Comment by paul-buerkner** [comment]
implemented via #577
